### PR TITLE
Bug fixes for RRFS 32-bit physics support, and minor feature additions

### LINKED ISSUE #1796: Add compilation lines without updating every `COMPILE` in the rt.conf
## Description

Presently, if you want to add a new `COMPILE` line to rt.conf, you have to increment the numbers of every COMPILE line after that.

## Solution
Use a name instead of a number:

```bash
COMPILE | p8_s2swa_dyn32 | intel | -DAPP=S2SWA -D32BIT=ON -DCCPP_SUITES=FV3_GFS_v17_coupled_p8,FV3_GFS_cpld_rasmgshocnsstnoahmp_ugwp | | fv3 |
RUN | cpld_control_p8_mixedmode                         |                            | baseline |
...

COMPILE | p8_s2swa | intel | -DAPP=S2SWA -DCCPP_SUITES=FV3_GFS_v17_coupled_p8,FV3_GFS_cpld_rasmgshocnsstnoahmp_ugwp | | fv3 |
RUN | cpld_control_p8                                   |                            | baseline |
...

COMPILE | p8_s2swa | gnu | -DAPP=S2SWA -DCCPP_SUITES=FV3_GFS_v17_coupled_p8,FV3_GFS_cpld_rasmgshocnsstnoahmp_ugwp | + hera cheyenne | fv3 |
RUN | cpld_control_p8                                   | + hera cheyenne            | baseline |
...

COMPILE | p8_s2s | gnu | -DAPP=S2S -DCCPP_SUITES=FV3_GFS_v17_coupled_p8,FV3_GFS_cpld_rasmgshocnsstnoahmp_ugwp | + hera cheyenne | fv3 |
RUN | cpld_control_nowave_noaero_p8                     | + hera cheyenne            | baseline |
```

The COMPILE_NR is the second column, then an _, and then the third column:

- `COMPILE | p8_s2swa | intel` => `p8_s2swa_intel`
- `COMPILE | p8_s2swa | gnu` => `p8_s2swa_gnu`

You can detect duplicated compile lines using gnu bash dictionaries, which requires bash 4.

This is implemented in #1797

## Alternatives
1. Remove the number entirely.
2. Step by 10 instead of 1. This way, nine compilations can be inserted between two of them before renumbering the file.

## Related to
Nothing.

### PR DESCRIPTION
**Pull request by SamuelTrahanNOAA**
## Description

Fixes two bugs and adds features:

1. aerinterp.F90 in CCPP stepped all over memory when kind_phys=4 because it read 8-byte reals into that array.
2. aerinterp.F90 did not report NetCDF errors up the call tree via errmsg & errflg
3. add regression tests for rrfs conus13km 32-bit physics and `-DFASTER=ON`
4. Initialize some uninitialized variables in GFS_typedefs.F90
5. Changes the `COMPILE` lines to use names instead of numbers so we don't have to renumber the entire rt.conf when adding a compilation line

### Input data additions/changes
- [x] No changes are expected to input data.

### Anticipated changes to regression tests:
- [x] Changes are expected to the following tests
New regression tests for 32-bit physics in RRFS

## Subcomponents involved:
- [ ] AQM
- [ ] CDEPS
- [ ] CICE
- [ ] CMEPS
- [ ] CMakeModules
- [x] FV3
- [ ] GOCART
- [ ] HYCOM
- [ ] MOM6
- [ ] NOAHMP
- [ ] WW3
- [ ] stochastic_physics
- [ ] none

## Commit Queue Checklist:

- [x] Link PR's from all sub-components involved in section below
- [x] Confirm reviews completed in ALL sub-component PR's
- [x] Add all appropriate labels to this PR.
- [x] Run full RT suite on either Hera/Cheyenne AND attach log to a PR comment.
- [x] Add list of any failed regression tests to "Anticipated changes to regression tests" section.

## Linked PR's and Issues:

- fixes #1796 
- depends on fv3atm https://github.com/NOAA-EMC/fv3atm/pull/668
- depends of ccpp-physics https://github.com/ufs-community/ccpp-physics/pull/85
- https://github.com/NOAA-EMC/fv3atm/issues/667
- https://github.com/ufs-community/ccpp-physics/issues/83
- https://github.com/ufs-community/ccpp-physics/issues/84

## Testing Day Checklist:

- [x] This PR is up-to-date with the top of all sub-component repositories except for those sub-components which are the subject of this PR.
- [ ] Move new/updated input data on RDHPCS Hera and propagate input data changes to all supported systems.

### Testing Log (for CM's):
- RDHPCS
  - [x] Hera
  - [x] Orion
  - [x] Jet
  - [x] Gaea
  - [ ] Cheyenne
- WCOSS2
  - [x] Dogwood/Cactus (IKE Day, skipping)
  - [x] Acorn
- CI
  - [x] Completed
- opnReqTest
  - [ ] N/A
  - [x] Log attached to comment

### CODE REVIEW

File: tests/rt.conf
Comment (DeniseWorthen) [inline tests/rt.conf]: I'm not sure about some of your name choices. I would never associate a compile called "p8_faster" with the fully coupled s2swa compile. Or p8_pdlib with the s2sw using unstructured ww3 meshes.

If I understand, the tests directory will contain exe files that are named ``p8_faster.exe`` and ``p8_pdlib.exe``? Or will they be ``fv3_p8_faster.exe``?

Comment (SamuelTrahanNOAA) [reply tests/rt.conf]: No, it'll be `fv3_p8_faster_intel.exe`. I don't know what some of these executables are. If you can suggest better names, I'd appreciate it.

File: tests/rt.conf
Comment (DeniseWorthen) [inline tests/rt.conf]: I would suggest ``p8_s2sw_pdlib`` and similar for the debug.

Comment (SamuelTrahanNOAA) [reply tests/rt.conf]: Done.

File: tests/rt.conf
Comment (DeniseWorthen) [reply tests/rt.conf]: You've made reasonable choices for the most part. But instead of ``p8_faster``, I'd use ``p8_s2swa_faster``.

I think instead of ``p8_s2s_cmeps``, I'd call it ``p8_s2s_aoflux``

File: tests/rt.conf
Comment (DeniseWorthen) [inline tests/rt.conf]: This is the same application as what you've called above just ``datm_cdeps``. It's definitely not ``p8`` so I'd go w/ the intel version.

File: tests/rt.conf
Comment (DeniseWorthen) [inline tests/rt.conf]: These two ``pdlib`` cases should have the same name as the intel compiles.

File: tests/rt.conf
Comment (SamuelTrahanNOAA) [reply tests/rt.conf]: Done.

File: tests/rt.conf
Comment (SamuelTrahanNOAA) [reply tests/rt.conf]: They match now.

File: tests/rt.conf
Comment (SamuelTrahanNOAA) [reply tests/rt.conf]: I've made these changes, but I'm leaving this conversation unresolved since you asked other questions.

File: tests/rt.conf
Comment (DeniseWorthen) [inline tests/rt.conf]: This one is tricky. For intel, this is called ``p8_s2sw``; it doesn't contain aerosols but we run one test w/ and one test w/o waves by turning off waves in the namelist. I think it is OK as-is.

File: tests/rt.conf
Comment (DeniseWorthen) [inline tests/rt.conf]: This is picky---can you make this `p8_s2swa_faster``? I might have mis-typed my earlier comment.

Comment (SamuelTrahanNOAA) [reply tests/rt.conf]: Done.

File: tests/rt.conf
Comment (DusanJovic-NOAA) [reply tests/rt.conf]: Not really important, but I think the first part of 'compile' name, which will be later used to generate the executable name, should be the 'APP' name, then other build options, maybe physics, or debug or 32bit etc. So in this case instead of p8_s2s_aoflux I would use s2s_p8_aoflux. It seems more logical to me. 
Also at some point we should rename all executable to start with `ufs_` instead `fv3_`, but I think we discussed this is some other issue or PR.

Comment (SamuelTrahanNOAA) [reply tests/rt.conf]: I just changed it to use the app name as the first part of the compilation name. Also, I got rid of "p8" since it was everywhere, and that seems redundant.

**Comment by jkbk2004** [comment]
@BrianCurtis-NOAA can I see if you agree with the replacement of compile exe numbers with more realistic application names ? Looks good to me.

**Comment by BrianCurtis-NOAA** [comment]
> @BrianCurtis-NOAA can I see if you agree with the replacement of compile exe numbers with more realistic application names ? Looks good to me.

A few things. opnReqTests might also have to be updated. AutoRT might also have to be updated. This meaning there are potential changes to be made in Jenkins.

I like the idea of the names, it really helps.

I also think we need to address the rapidly growing number of tests being added to UFS RT. But not in this PR.

**Comment by jkbk2004** [comment]
@SamuelTrahanNOAA Can you sync up branch? So we may start some sanity check. If you have a regression test log, please go ahead to attach.

**Comment by SamuelTrahanNOAA** [comment]
I will sync up the branch. I have no recent regression test logs, so I'll have to rerun the regression tests.

**Comment by SamuelTrahanNOAA** [comment]
Merging the rt.conf changes were complicated due to the abundance of conflicts. I definitely need to rerun the regression tests before this is merged.

**Comment by SamuelTrahanNOAA** [comment]
I can confirm that all tests pass except my new ones on Hera. I'm going to run my new regression tests next.

**Comment by SamuelTrahanNOAA** [comment]
Everything passed, and I uploaded the hera logs.

**Comment by SamuelTrahanNOAA** [comment]
The baseline I compared against was just the emc.nemspara baseline, with my new tests added in. I only generated the baselines for those new tests; all others were the original.

File: tests/rt.conf
Comment (BrianCurtis-NOAA) [inline tests/rt.conf:1]: Line 4 (item 2) should be adjusted with your changes.

**Review by BrianCurtis-NOAA — CHANGES_REQUESTED** [review]

File: tests/rt.conf
Comment (BrianCurtis-NOAA) [inline tests/rt.conf:401]: There is already a s2swa in the "Intel" section, I think this will try to use compile_swswa directory and if there is already one, will cause complications if they are running at the same time. If there are similar compile names maybe add _gnu to the GNU compiles?

**Review by BrianCurtis-NOAA — CHANGES_REQUESTED** [review]

File: tests/rt.conf
Comment (SamuelTrahanNOAA) [reply tests/rt.conf:401]: It already adds _gnu or _intel to the end of the compile_nr. If you type `s2swa`, you get `s2swa_gnu` or `s2swa_intel`.

Comment (SamuelTrahanNOAA) [reply tests/rt.conf:401]: Also, it checks for duplicate compilation names. It uses hashes, which require bash 4.2 or later. Older bash versions will abort due to syntax errors.

File: tests/rt.conf
Comment (SamuelTrahanNOAA) [reply tests/rt.conf:1]: I fixed this already. Perhaps you were looking at an older version?

File: tests/rt.conf
Comment (BrianCurtis-NOAA) [reply tests/rt.conf:401]: OK, sounds good to me. Thanks for the clarification.

File: tests/rt.conf
Comment (BrianCurtis-NOAA) [reply tests/rt.conf:1]: This?: https://github.com/ufs-community/ufs-weather-model/blob/86212e1656fc89ee323ccccfb92d3a17a91028a4/tests/rt.conf#L4 should show as compile_name with these changes, right?

Comment (SamuelTrahanNOAA) [reply tests/rt.conf:1]: I just updated that documentation. Sorry, I thought you were referring to `COMPILE | 4 |`.

Comment (SamuelTrahanNOAA) [reply tests/rt.conf:1]: > Item 2: Compile name - The rt.sh will add _gnu or _intel to it. There must be no duplicate names for each compiler.
>                        If two compilations are identical except compiler, please use the same name for each.

**Comment by jkbk2004** [comment]
@SamuelTrahanNOAA We can start working on this pr. Can you sync up branches? Also, #1810 is ready to be combined. Can you go ahead to merge in #1810 ?

**Comment by SamuelTrahanNOAA** [comment]
> We can start working on this pr. Can you sync up branches? Also, https://github.com/ufs-community/ufs-weather-model/pull/1810 is ready to be combined. Can you go ahead to merge in https://github.com/ufs-community/ufs-weather-model/pull/1810 ?

Okay. I'm doing it now.

**Comment by SamuelTrahanNOAA** [comment]
The branch is updated, and I've merged #1810. I want to test this combined version to make sure Rocoto support still works.

**Comment by DeniseWorthen** [comment]
Can I ask that you hold off for a bit---I have seen an issue w/ the PR I just merged (evap from atm). Let me investigate.

**Comment by SamuelTrahanNOAA** [comment]
> Can I ask that you hold off for a bit---I have seen an issue w/ the PR I just merged (evap from atm). Let me investigate.

I already merged develop and #1810 to this branch.

**Comment by DeniseWorthen** [comment]
Yes, but the issue I found may change answers. I realize you're not testing coupled so it most likely won't impact you, if you're comfortable w/ that caveat.

**Comment by SamuelTrahanNOAA** [comment]
> Yes, but the issue I found may change answers. I realize you're not testing coupled so it most likely won't impact you, if you're comfortable w/ that caveat.

I'm running the full regression test suite on Hera.

**Comment by SamuelTrahanNOAA** [comment]
The changes in #1810 broke the `rt.sh -m` option. Those changes moved the getopts loop up, to before most variables were set. Thus, the assignments, like `RTPWD=${NEW_BASELINE}`, access uninitialized variables.

**Comment by BrianCurtis-NOAA** [comment]
> The changes in #1810 broke the `rt.sh -m` option. Those changes moved the getopts loop up, to before most variables were set. Thus, the assignments, like `RTPWD=${NEW_BASELINE}`, access uninitialized variables.

OK revert that. I'll need to do a deeper dive into things. Good catch Sam.

**Comment by SamuelTrahanNOAA** [comment]
> OK revert that. I'll need to do a deeper dive into things. Good catch Sam.

I have a fix that I'm testing now.

**Comment by SamuelTrahanNOAA** [comment]
I've never used the `rt.sh -n` option. Was that known to work before these changes?

The only experience I had was someone telling me a test failed with `-n` while it succeeded without.

**Comment by DeniseWorthen** [comment]
OK, false (ish) alarm. There is definitely an issue in the evap export, but it only changes the fill value. I had a panic moment because we had an hrrr control test fail to reproduce baselines on WCOSS2. I couldn't imagine it was related, but wanted to check.

**Comment by BrianCurtis-NOAA** [comment]
> I've never used the `rt.sh -n` option. Was that known to work before these changes?
> 
> The only experience I had was someone telling me a test failed with `-n` while it succeeded without.

Yes, -n just runs the one test, you can use it with say control_p8 Intel for your testing.

**Comment by SamuelTrahanNOAA** [comment]
I have fixes that let `-c`, `-n`, `-m`, and `-l` work in Rocoto mode (`-r`). I'm running the full regression test suite on Hera now. Once it passes, I'll push my fixes and the log file. It is 60% of the way through the tests, and I'm running 40 jobs at a time. It should be done quite soon.

**Comment by BrianCurtis-NOAA** [comment]
> I have fixes that let `-c`, `-n`, `-m`, and `-l` work in Rocoto mode (`-r`). I'm running the full regression test suite on Hera now. Once it passes, I'll push my fixes and the log file. It is 60% of the way through the tests, and I'm running 40 jobs at a time. It should be done quite soon.

I assume -e still works with your changes.

**Comment by SamuelTrahanNOAA** [comment]
> I assume -e still works with your changes.

It worked before I merged #1810. The ecFlow system should get thorough exercise when you run the regression tests everywhere.

**Comment by SamuelTrahanNOAA** [comment]
My tests passed, so it is time for you to run yours.

**Comment by SamuelTrahanNOAA** [comment]
@jkbk2004 - The baseline doesn't change. Only the new tests have new results. It is important that you do not regenerate the baselines for the other tests. That will let us confirm nothing else has broken.

**Comment by jkbk2004** [comment]
@SamuelTrahanNOAA Thanks for reminding!

**Comment by zach1221** [comment]
Jenkins-ci log attached. ORTs all passed.
[ufs-weather-model » ort-docker-pipeline » PR-1797 #2 Console [Jenkins].pdf](https://github.com/ufs-community/ufs-weather-model/files/11991105/ufs-weather-model.ort-docker-pipeline.PR-1797.2.Console.Jenkins.pdf)

**Comment by SamuelTrahanNOAA** [comment]
> Jenkins-ci log attached. ORTs all passed.

Why does github say the ORT tests are still "pending - building stage?"

I changed the rt.conf compile lines to use names, not numbers, so it may have broken your system.

**Comment by zach1221** [comment]
@SamuelTrahanNOAA I'm not certain, as the job is listed as complete in Jenkins. I can investigate to find out.

File: tests/tests/rrfs_conus13km_phy32_qr
Comment (BrianCurtis-NOAA) [inline tests/tests/rrfs_conus13km_phy32_qr]: I notice this is a new test, but it is not in rt.conf. Is this intended?

**Comment by BrianCurtis-NOAA** [comment]
Also, have you tested this using -a? Acorn is giving me an error using -a like it's never being set: "Please use -a <account> to set group account to use on HPC" but I am using it.

**Comment by junwang-noaa** [comment]
@SamuelTrahanNOAA is it true that currently only "FV3_HRRR" ccpp physics package can run with 32-bit? Any other suite file can also run with 32-bit too?

File: tests/rt.sh
Comment (BrianCurtis-NOAA) [inline tests/rt.sh]: Having this before the while getopts will cause anyone who uses -a to fail. It needs to go directly after the while getopts to work. Why was this part moved before?

**Review by BrianCurtis-NOAA — CHANGES_REQUESTED** [review]

**Comment by zach1221** [comment]
@BrianCurtis-NOAA I had the same issue, and found hashing out lines 135 through 138 in rt.sh worked. However this is probably not the preferred fix. Otherwise I think you can export ACCNR prior to running.

**Comment by BrianCurtis-NOAA** [comment]
> @BrianCurtis-NOAA I had the same issue, and found hashing out lines 135 through 138 in rt.sh worked. However this is probably not the preferred fix. Otherwise I think you can export ACCNR prior to running.

Are you editing code during testing to get it working?

If you are, it needs to be brought back into the PR, discussed, or not changed at all.

**Comment by jkbk2004** [comment]
@SamuelTrahanNOAA An option might be : move lines 135-138 after line 453 and keep machine ACCNR default setting along with if blocks around DISKNM setting lines.

**Comment by BrianCurtis-NOAA** [comment]
> @SamuelTrahanNOAA An option might be : move lines 135-138 after line 453 and keep machine ACCNR default setting along with if blocks around DISKNM setting lines.

line 216 would be best and all lines noted above.

**Comment by jkbk2004** [comment]
Otherwise users just need "export ACCNR=###" before running rt.

**Comment by BrianCurtis-NOAA** [comment]
> Otherwise users just need "export ACCNR=###" before running rt.

Right, but we _NEED_ ACCNR set and -a forces the user to set it if they don't have ACCNR set, and thus -a needs to work properly.

**Comment by DeniseWorthen** [comment]
> > @BrianCurtis-NOAA I had the same issue, and found hashing out lines 135 through 138 in rt.sh worked. However this is probably not the preferred fix. Otherwise I think you can export ACCNR prior to running.
> 
> Are you editing code during testing to get it working?
> 
> If you are, it needs to be brought back into the PR, discussed, or not changed at all.

We need to understand---are the RTs being tested with uncommitted changes?

**Comment by jkbk2004** [comment]
I agree with forcing along with line 216 in getopts block.

**Comment by zach1221** [comment]
@DeniseWorthen No, sorry. To clarify you can use the export command, for example "export ACCNR=${ACCNR:-epic}", on Orion, before running ./rt.sh. It doesn't require any changes to the rt.sh script. This is how I ran the tests, but I thought I'd offer the other workaround as well, while acknowledging it's likely not preferred.

**Comment by DeniseWorthen** [comment]
@zach1221 @jkbk2004 Have any RT logs been posted for cases where the PR was run w/o committed changes?

Edit: Thanks for the clarification.

**Comment by SamuelTrahanNOAA** [comment]
Do people want me to make changes or not? If so, what are the changes? I'm seeing too many conflicting comments.

**Comment by jkbk2004** [comment]
> Do people want me to make changes or not? If so, what are the changes? I'm seeing too many conflicting comments.

@SamuelTrahanNOAA Please, go ahead to add fix. Sounds like option to force somewhere inside getopts block

**Comment by BrianCurtis-NOAA** [comment]
> Do people want me to make changes or not? If so, what are the changes? I'm seeing too many conflicting comments.

Move lines:  +134 to +141 to 216 (after the while getopts section)

**Comment by SamuelTrahanNOAA** [comment]
> Move lines: +134 to +141 to 216 (after the while getopts section)

I tested that fix, and it works. With `-a`, the workflow uses the account I give it. Without -a, it aborts with the expected message.

The fix is in my branch now.

File: tests/rt.sh
Comment (BrianCurtis-NOAA) [reply tests/rt.sh]: Thanks!

File: tests/tests/rrfs_conus13km_phy32_qr
Comment (BrianCurtis-NOAA) [reply tests/tests/rrfs_conus13km_phy32_qr]: If you want that test enabled, we can do that in a future PR.

File: tests/rt.conf
Comment (BrianCurtis-NOAA) [reply tests/rt.conf]: @DusanJovic-NOAA @DeniseWorthen are you satisfied with this solution? Trying to get these issues resolved as we finish up testing.

File: tests/tests/rrfs_conus13km_phy32_qr
Comment (SamuelTrahanNOAA) [reply tests/tests/rrfs_conus13km_phy32_qr]: I just removed it.

File: tests/rt.conf
Comment (DeniseWorthen) [reply tests/rt.conf]: Yes, I'm OK w/ the new names now.

Comment (DusanJovic-NOAA) [reply tests/rt.conf]: :+1:

**Review by BrianCurtis-NOAA — DISMISSED** [review]

**Comment by BrianCurtis-NOAA** [comment]
@jkbk2004 @zach1221 other than Cheyenne, give it a look over and I assume you can start the merging.

**Comment by zach1221** [comment]
Thanks, @BrianCurtis-NOAA . Yes, since we're skipping Cheyenne, testing is complete and we can begin the merging process.

**Comment by dustinswales** [comment]
@zach1221 Physics has been merged.
@SamuelTrahanNOAA You can go ahead and update your FV3 submodule pointer.

**Comment by SamuelTrahanNOAA** [comment]
> You can go ahead and update your FV3 submodule pointer.

I updated the fv3atm's submodule pointer to ccpp-physics. It now points to the authoritative ufs/dev branch. You can merge the fv3atm PR.

**Comment by zach1221** [comment]
FV3 sub-pr is merged. Please update the submodule pointer and revert the .gitmodules for FV3. @SamuelTrahanNOAA

**Comment by SamuelTrahanNOAA** [comment]
> FV3 sub-pr is merged. Please update the submodule pointer and revert the .gitmodules for FV3

The deed is done. We are ready for final approvals and a merge.

**Review by BrianCurtis-NOAA — APPROVED** [review]

**Review by jkbk2004 — APPROVED** [review]

**Pull request merged**
